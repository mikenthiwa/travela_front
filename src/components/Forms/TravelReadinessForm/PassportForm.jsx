import React from 'react';
import './TravelReadiness.scss';
import withTravelReadinessForm from '../../Hoc/withTravelReadinessForm';
import PassportDetailsFieldSet from './FormFieldsets/passportDetails';

export default withTravelReadinessForm(PassportDetailsFieldSet,'passport',
  {
    values: {
      name: '',
      passportNumber: '',
      nationality:'',
      dateOfBirth: '',
      dateOfIssue: '',
      placeOfIssue: '',
      expiryDate: '',
    },
   
    image: '',
    imageChanged: false,
    documentUploaded: false,
    uploadingDocument: false,
    errors: {},
    hasBlankFields: true
  });

