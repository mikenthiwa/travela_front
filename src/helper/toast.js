import toast from 'toastr';
import cancelBtn from '../images/toast-cancel.svg';

toast.options = {
  progressBar: false,
  closeButton: true,
  preventDuplicates: true,
  positionClass: 'toast-top-center',
  closeHtml: `
    <button class="custom-close-btn">
      <img alt="close btn"/>
    </button>
  `,
};

export const successMessage = message =>  toast.success(message);
export const errorMessage = message =>  toast.error(message);


