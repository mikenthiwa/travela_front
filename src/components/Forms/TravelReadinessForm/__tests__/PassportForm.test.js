import toast from 'toastr';
import moxios from 'moxios';
import React from 'react';
import moment from 'moment';
import PassportForm from '../PassportForm';
import PassportDetailsFieldSet from '../FormFieldsets/passportDetails';

toast.success = jest.fn();
toast.error = jest.fn();

describe('<PassportForm/>',  () => {

  let wrapper;

  const props = {
    createTravelReadinessDocument: jest.fn(),
    editTravelReadinessDocument: jest.fn(),
    scaningPassport: jest.fn(),
    fetchingDocument: false,
    isLoading: false,
    document: { },
    errors: {},
    fetchUserData: jest.fn(),
    users: [],
    user: {},
    closeModal: jest.fn(),
    onCancel: jest.fn(),
    passportInfo:{},
    modalType:'Add Passport',
    scanning:false,
    showPassportForm:true
  };

  const document = {
    name: 'Mike',
    passportNumber : 'ABC123aemlbczy85',
    nationality : 'Kenyan ',
    placeOfIssue : 'Kenya',
    cloudinaryUrl : 'https://farm4.staticflickr.com/3894/15008518202_c265dfa55f_h.jpeg'
  };


  const passportFile = new Blob(['Passport file'], {type: 'application/msword', name: 'Filename.doc'});

  const event = {
    target: {
      files: [
        passportFile
      ]
    }
  };

  beforeEach(() => {
    wrapper = mount(<PassportForm {...props} />);
    moxios.install();
  });

  afterEach(() => {
    moxios.uninstall();
    jest.resetAllMocks();
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the edit passport document modal', () => {
    const wrapperWithPassportField = mount(
      <PassportForm
        {...props}
        documentType="passport"
        modalType="edit passport"
      />
    );
    expect(
      wrapperWithPassportField.find(PassportDetailsFieldSet))
      .toHaveLength(1);
  });

  it('renders passport  modal with button text \'Add Passport\'', () => {
    const wrapperWithOtherDocumentField = mount(
      <PassportForm {...props} documentType="passport" />
    );
    expect(
      wrapperWithOtherDocumentField.find('button#submit').text()).toEqual('Add Passport');
  });

  it('renders edit passport  modal with button text \'Save Changes\'', () => {
    const wrapperWithOtherDocumentField = mount(
      <PassportForm {...props} documentType="passport" modalType="edit passport" />
    );
    expect(
      wrapperWithOtherDocumentField.find('button#submit').text()).toEqual('Save Changes');
  });

  it('should not submit without document', () => {
    wrapper.setState({values: {...document}});
    wrapper.setProps({document: {data: { imageName: 'image.jpg' }}});
    process.env.REACT_APP_CLOUNDINARY_API = 'https://mock-passport-cloudinary-api-succeed';

    const cloudinaryResponse = {
      url: 'secure url'
    };

    moxios.stubRequest(process.env.REACT_APP_CLOUNDINARY_API, {
      status: 200,
      response: cloudinaryResponse
    });

    wrapper.find('form').simulate('submit');

    moxios.wait(() => {
      expect(wrapper.state().documentUploaded).toBeFalsy();
      expect(toast.error).toHaveBeenCalledWith('Please upload a document.');
    });
  });


  it('should submit if all the data is valid',  () => {
    wrapper.setState({values: {...document}});
    wrapper.find('#select-file').simulate('change', event);
    wrapper.setProps({document: {data: { imageName: 'image.jpg' }}});

    process.env.REACT_APP_CLOUNDINARY_API = 'https://mock-passport-cloudinary-api-succeed';

    const cloudinaryResponse = {
      url: 'secure url'
    };

    moxios.stubRequest(process.env.REACT_APP_CLOUNDINARY_API, {
      status: 200,
      response: cloudinaryResponse
    });

    wrapper.find('form').simulate('submit');

    moxios.wait(() => {
      expect(wrapper.state().documentUploaded).toBeTruthy();
      expect(props.createTravelReadinessDocument).toHaveBeenCalledWith('passport', document);
    });
  });

  it('should toast an error if the passport file does not upload', () => {
    wrapper.setState({values: {...document}});
    wrapper.find('#select-file').simulate('change', event);
    wrapper.setProps({document: {data: { imageName: 'image.jpg' }}});

    process.env.REACT_APP_CLOUNDINARY_API = 'https://mock-passport-cloudinary-api-succeed';

    const cloudinaryResponse = {
      url: 'secure url'
    };

    moxios.stubRequest(process.env.REACT_APP_CLOUNDINARY_API, {
      status: 404,
      response: cloudinaryResponse
    });

    wrapper.setState({documentUploaded: false});
    wrapper.find('form').simulate('submit');
    expect(toast.error).toHaveBeenCalledWith('Invalid file type. Please upload an image');
  });

  it('toasts an error if cloudinary returns an error', () => {
    moxios.stubRequest(process.env.REACT_APP_CLOUNDINARY_API, { status: 500});
    wrapper.find('input[type="file"]').simulate('change', event);
  });

  it('should edit if all the data is valid',  () => {
    const newProps = {
      ...props,
      modalType: 'edit passport'
    };

    wrapper = mount(<PassportForm {...newProps} />);
    wrapper.setState({values: {...document}});
    wrapper.find('input[type="file"]').simulate('change', event);
    wrapper.setProps({document: {data: { imageName: 'image.jpg' }}});

    process.env.REACT_APP_CLOUNDINARY_API = 'https://mock-passport-cloudinary-api-succeed';

    const cloudinaryResponse = {
      url: 'secure url'
    };

    moxios.stubRequest(process.env.REACT_APP_CLOUNDINARY_API, {
      status: 200,
      response: cloudinaryResponse
    });

    wrapper.find('form').simulate('submit');

    moxios.wait(() => {
      expect(wrapper.state().documentUploaded).toBeTruthy();
      expect(props.createTravelReadinessDocument).toHaveBeenCalledWith('passport', document);
    });
  });


  it('should display an error when there is no value in input', () => {
    const formWrapper = mount(
      <PassportForm {...props} documentType="passport" />
    );
    formWrapper.setProps({document: {data: { imageName: 'image.jpg' }}});
    let documentUploaded = false;
    formWrapper.find('form').simulate('submit', {
      preventDefault: () => {
        documentUploaded = true;
      },
    });
    expect(documentUploaded).toBe(true);
  });

  it('should receive new props for the input fields if user wants to edit a document', () => {
    const newProps = {
      ...props,
      editTravelReadinessDocument: jest.fn(),
      document: {
        data: {
          passportNumber: '6yy2',
          name: 'emeka',
          nationality: 'Nigerian',
          placeOfIssue: 'Naija',
          cloudinaryUrl: 'image.com'
        }
      },
      currentDocument: {},
    };
    wrapper = mount(<PassportForm {...newProps} />);
    wrapper.setState({values: {...document}});
  });

  it('should cancel if user no longer wants to edit a document', () => {
    const newProps = {
      ...props,
      editTravelReadinessDocument: jest.fn(),
      document: {
        data: {
          passportNumber: '6yy2',
          name: 'emeka',
          nationality: 'Nigerian',
          placeOfIssue: 'Naija',
          cloudinaryUrl: 'image.com'
        }
      },
      currentDocument: {},
    };
    wrapper = mount(<PassportForm {...newProps} />);
    wrapper.setState({values: {...document}});
  });

  it('should show the year from 18 years ago on the birthdate', () => {
    wrapper = mount(<PassportForm {...props} />);
    const datePicker = wrapper.find('DateInput[name="dateOfBirth"]').find('DatePicker');
    expect(datePicker.props().showYearDropdown).toBeTruthy();
    expect(moment(datePicker.props().maxDate).year()).toEqual(moment().subtract(18,'year').year());
  });
});
