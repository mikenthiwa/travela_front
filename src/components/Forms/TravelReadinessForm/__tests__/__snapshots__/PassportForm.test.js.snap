// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PassportForm/> renders correctly 1`] = `
<div>
  <form
    className="travel-document-form"
    key=".0"
    onSubmit={[Function]}
  >
    <div
      className="travel-document-select-file"
    >
      <p
        className="attach-document-text"
      >
        Attach the image or PDF of your  document
      </p>
      <div
        className="document-input__input-container"
      >
        <label
          className="document-input__input-container__prompts"
          htmlFor="select-file"
        >
          <div
            className="document-input__input-container__prompts__img"
          >
            <img
              alt="Document Upload"
              src="document-upload-blue.svg"
            />
          </div>
          <div
            className="document-input__input-container__prompts__text"
          >
            <p>
              Choose from computer
            </p>
            <p
              className="maximum-file-size"
            >
              Maximum file size - 10MB
            </p>
          </div>
          <input
            accept="image/jpg, image/png, application/pdf"
            id="select-file"
            onChange={[Function]}
            type="file"
          />
        </label>
      </div>
    </div>
    <div>
      <fieldset>
        <div
          className="input-group resposive_passport"
        >
          <div
            className="form-input fluid "
          >
            <label
              htmlFor="name"
            >
              Name as Seen On Passport
              <span
                style={
                  Object {
                    "color": "red",
                  }
                }
              >
                 * 
              </span>
            </label>
            <input
              autoComplete="off"
              className=" fluid"
              label="Name as Seen On Passport"
              name="name"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
            <span
              className="error"
            />
          </div>
          <div
            className="form-input  "
          >
            <label
              htmlFor="passportNumber"
            >
              Passport Number
              <span
                style={
                  Object {
                    "color": "red",
                  }
                }
              >
                 * 
              </span>
            </label>
            <input
              autoComplete="off"
              className=" "
              label="Passport Number"
              name="passportNumber"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
            <span
              className="error"
            />
          </div>
          <div
            className="form-input  "
          >
            <label
              htmlFor="nationality"
            >
              Nationality
              <span
                style={
                  Object {
                    "color": "red",
                  }
                }
              >
                 * 
              </span>
            </label>
            <div
              style={
                Object {
                  "position": "relative",
                }
              }
            >
              <div
                className="value"
                style={
                  Object {
                    "width": "",
                  }
                }
              >
                <input
                  autoComplete="off"
                  choices={
                    Array [
                      "Aruban",
                      "Afghan",
                      "Angolan",
                      "Anguillian",
                      "Ålandish",
                      "Albanian",
                      "Andorran",
                      "Emirati",
                      "Argentine",
                      "Armenian",
                      "American Samoan",
                      "Antarctican",
                      "French",
                      "Antiguan, Barbudan",
                      "Australian",
                      "Austrian",
                      "Azerbaijani",
                      "Burundian",
                      "Belgian",
                      "Beninese",
                      "Burkinabe",
                      "Bangladeshi",
                      "Bulgarian",
                      "Bahraini",
                      "Bahamian",
                      "Bosnian, Herzegovinian",
                      "Saint Barthélemy Islander",
                      "Saint Helenian",
                      "Belarusian",
                      "Belizean",
                      "Bermudian",
                      "Bolivian",
                      "Dutch",
                      "Brazilian",
                      "Barbadian",
                      "Bruneian",
                      "Bhutanese",
                      "",
                      "Motswana",
                      "Central African",
                      "Canadian",
                      "Cocos Islander",
                      "Swiss",
                      "Chilean",
                      "Chinese",
                      "Ivorian",
                      "Cameroonian",
                      "Congolese",
                      "Cook Islander",
                      "Colombian",
                      "Comoran",
                      "Cape Verdian",
                      "Costa Rican",
                      "Cuban",
                      "Christmas Islander",
                      "Caymanian",
                      "Cypriot",
                      "Czech",
                      "German",
                      "Djibouti",
                      "Dominican",
                      "Danish",
                      "Algerian",
                      "Ecuadorean",
                      "Egyptian",
                      "Eritrean",
                      "Sahrawi",
                      "Spanish",
                      "Estonian",
                      "Ethiopian",
                      "Finnish",
                      "Fijian",
                      "Falkland Islander",
                      "Faroese",
                      "Micronesian",
                      "Gabonese",
                      "British",
                      "Georgian",
                      "Channel Islander",
                      "Ghanaian",
                      "Gibraltar",
                      "Guinean",
                      "Guadeloupian",
                      "Gambian",
                      "Guinea-Bissauan",
                      "Equatorial Guinean",
                      "Greek",
                      "Grenadian",
                      "Greenlandic",
                      "Guatemalan",
                      "Guianan",
                      "Guamanian",
                      "Guyanese",
                      "Hong Konger",
                      "Heard and McDonald Islander",
                      "Honduran",
                      "Croatian",
                      "Haitian",
                      "Hungarian",
                      "Indonesian",
                      "Manx",
                      "Indian",
                      "Irish",
                      "Iranian",
                      "Iraqi",
                      "Icelander",
                      "Israeli",
                      "Italian",
                      "Jamaican",
                      "Jordanian",
                      "Japanese",
                      "Kazakhstani",
                      "Kenyan",
                      "Kirghiz",
                      "Cambodian",
                      "I-Kiribati",
                      "Kittitian or Nevisian",
                      "South Korean",
                      "Kosovar",
                      "Kuwaiti",
                      "Laotian",
                      "Lebanese",
                      "Liberian",
                      "Libyan",
                      "Saint Lucian",
                      "Liechtensteiner",
                      "Sri Lankan",
                      "Mosotho",
                      "Lithuanian",
                      "Luxembourger",
                      "Latvian",
                      "Macanese",
                      "Saint Martin Islander",
                      "Moroccan",
                      "Monegasque",
                      "Moldovan",
                      "Malagasy",
                      "Maldivan",
                      "Mexican",
                      "Marshallese",
                      "Macedonian",
                      "Malian",
                      "Maltese",
                      "Bamar",
                      "Montenegrin",
                      "Mongolian",
                      "American",
                      "Mozambican",
                      "Mauritanian",
                      "Montserratian",
                      "Mauritian",
                      "Malawian",
                      "Malaysian",
                      "Mahoran",
                      "Namibian",
                      "New Caledonian",
                      "Nigerien",
                      "Norfolk Islander",
                      "Nigerian",
                      "Nicaraguan",
                      "Niuean",
                      "Norwegian",
                      "Nepalese",
                      "Nauruan",
                      "New Zealander",
                      "Omani",
                      "Pakistani",
                      "Panamanian",
                      "Pitcairn Islander",
                      "Peruvian",
                      "Filipino",
                      "Palauan",
                      "Papua New Guinean",
                      "Polish",
                      "Puerto Rican",
                      "North Korean",
                      "Portuguese",
                      "Paraguayan",
                      "Palestinian",
                      "French Polynesian",
                      "Qatari",
                      "Romanian",
                      "Russian",
                      "Rwandan",
                      "Saudi Arabian",
                      "Sudanese",
                      "Senegalese",
                      "Singaporean",
                      "South Georgian South Sandwich Islander",
                      "Solomon Islander",
                      "Sierra Leonean",
                      "Salvadoran",
                      "Sammarinese",
                      "Somali",
                      "Serbian",
                      "South Sudanese",
                      "Sao Tomean",
                      "Surinamer",
                      "Slovak",
                      "Slovene",
                      "Swedish",
                      "Swazi",
                      "St. Maartener",
                      "Seychellois",
                      "Syrian",
                      "Turks and Caicos Islander",
                      "Chadian",
                      "Togolese",
                      "Thai",
                      "Tadzhik",
                      "Tokelauan",
                      "Turkmen",
                      "East Timorese",
                      "Tongan",
                      "Trinidadian",
                      "Tunisian",
                      "Turkish",
                      "Tuvaluan",
                      "Taiwanese",
                      "Tanzanian",
                      "Ugandan",
                      "Ukrainian",
                      "Uruguayan",
                      "Uzbekistani",
                      "Vatican",
                      "Saint Vincentian",
                      "Venezuelan",
                      "Virgin Islander",
                      "Vietnamese",
                      "Ni-Vanuatu",
                      "Wallis and Futuna Islander",
                      "Samoan",
                      "Yemeni",
                      "South African",
                      "Zambian",
                      "Zimbabwean",
                    ]
                  }
                  className="occupationInput"
                  label="Nationality"
                  name="role"
                  onChange={[Function]}
                  size=""
                  type="text"
                  value=""
                />
              </div>
            </div>
            <span
              className="error"
            />
          </div>
          <div
            className="form-input  "
          >
            <label
              htmlFor="dateOfBirth"
            >
              Date of Birth (MM/DD/YYYY)
              <span
                style={
                  Object {
                    "color": "red",
                  }
                }
              >
                 * 
              </span>
            </label>
            <div
              className="date-wrapper    "
              id="dateOfBirth_date"
            >
              <div>
                <div
                  className="react-datepicker-wrapper"
                >
                  <div
                    className="react-datepicker__input-container"
                  >
                    <input
                      autoComplete="off"
                      className=""
                      disabled={false}
                      name="dateOfBirth"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onClick={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      placeholder="MM/DD/YYYY"
                      readOnly={false}
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
              <img
                alt="cal"
                className="calendar-icon"
                src="calendar_icon.svg"
              />
            </div>
            <span
              className="error"
            />
          </div>
          <div
            className="form-input  "
          >
            <label
              htmlFor="dateOfIssue"
            >
              Date of Issue (MM/DD/YYYY)
              <span
                style={
                  Object {
                    "color": "red",
                  }
                }
              >
                 * 
              </span>
            </label>
            <div
              className="date-wrapper    "
              id="dateOfIssue_date"
            >
              <div>
                <div
                  className="react-datepicker-wrapper"
                >
                  <div
                    className="react-datepicker__input-container"
                  >
                    <input
                      autoComplete="off"
                      className=""
                      disabled={false}
                      name="dateOfIssue"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onClick={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      placeholder="MM/DD/YYYY"
                      readOnly={false}
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
              <img
                alt="cal"
                className="calendar-icon"
                src="calendar_icon.svg"
              />
            </div>
            <span
              className="error"
            />
          </div>
          <div
            className="form-input  "
          >
            <label
              htmlFor="placeOfIssue"
            >
              Place of Issue
              <span
                style={
                  Object {
                    "color": "red",
                  }
                }
              >
                 * 
              </span>
            </label>
            <input
              autoComplete="off"
              className=" "
              label="Place of Issue"
              name="placeOfIssue"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
            <span
              className="error"
            />
          </div>
          <div
            className="form-input  "
          >
            <label
              htmlFor="expiryDate"
            >
              Expiry Date (MM/DD/YYYY)
              <span
                style={
                  Object {
                    "color": "red",
                  }
                }
              >
                 * 
              </span>
            </label>
            <div
              className="date-wrapper    "
              id="expiryDate_date"
            >
              <div>
                <div
                  className="react-datepicker-wrapper"
                >
                  <div
                    className="react-datepicker__input-container"
                  >
                    <input
                      autoComplete="off"
                      className=""
                      disabled={false}
                      name="expiryDate"
                      onBlur={[Function]}
                      onChange={[Function]}
                      onClick={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      placeholder="MM/DD/YYYY"
                      readOnly={false}
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
              <img
                alt="cal"
                className="calendar-icon"
                src="calendar_icon.svg"
              />
            </div>
            <span
              className="error"
            />
          </div>
        </div>
      </fieldset>
    </div>
    <hr />
    <div
      className="travel-document-submit-area"
    >
      <fieldset>
        <div
          className="submit-area"
        >
          <button
            className="bg-btn bg-btn--inactive"
            disabled={false}
            id="cancel"
            onClick={[Function]}
            type="button"
          >
            Cancel
          </button>
          <button
            className="bg-btn bg-btn--active"
            disabled={true}
            id="submit"
            type="submit"
          >
            <undefined />
          </button>
        </div>
      </fieldset>
    </div>
  </form>
</div>
`;
